examples/vision/semisupervised_simclr.py

```diff
@@ -0,0 +1,631 @@
+        similarities = (
+            tf.matmul(projections_1, projections_2, transpose_b=True) / self.temperature
+        )
+
+        # the similarity between the representations of two augmented views of the
+        # same image should be higher than their similarity with other views
+        contrastive_labels = tf.range(batch_size)
+        self.contrastive_accuracy.update_state(contrastive_labels, similarities)
+        self.contrastive_accuracy.update_state(
+            contrastive_labels, tf.transpose(similarities)
+        )
+
+        # the temperature-scaled similarities are used as logits for cross-entropy
+        # a symmetrized version of the loss is used here
+        loss_1_2 = keras.losses.sparse_categorical_crossentropy(
+            contrastive_labels, similarities, from_logits=True
+        )
+        loss_2_1 = keras.losses.sparse_categorical_crossentropy(
+            contrastive_labels, tf.transpose(similarities), from_logits=True
+        )
+        return (loss_1_2 + loss_2_1) / 2
+
+    def train_step(self, data):
+        (unlabeled_images, _), (labeled_images, labels) = data
+
+        # both labeled and unlabeled images are used, without labels
+        images = tf.concat((unlabeled_images, labeled_images), axis=0)
+        # each image is augmented twice, differently
+        augmented_images_1 = self.contrastive_augmenter(images)
+        augmented_images_2 = self.contrastive_augmenter(images)
+        with tf.GradientTape() as tape:
+            features_1 = self.encoder(augmented_images_1)
+            features_2 = self.encoder(augmented_images_2)
+            # the representations are passed through a projection mlp
+            projections_1 = self.projection_head(features_1)
+            projections_2 = self.projection_head(features_2)
+            contrastive_loss = self.contrastive_loss(projections_1, projections_2)
+        gradients = tape.gradient(
+            contrastive_loss,
+            self.encoder.trainable_weights + self.projection_head.trainable_weights,
+        )
+        self.contrastive_optimizer.apply_gradients(
+            zip(
+                gradients,
+                self.encoder.trainable_weights + self.projection_head.trainable_weights,
+            )
+        )
+
+        # labels are only used in evalutation for an on-the-fly logistic regression
+        preprocessed_images = self.classification_augmenter(labeled_images)
+        with tf.GradientTape() as tape:
+            features = self.encoder(preprocessed_images)
+            class_logits = self.linear_probe(features)
+            probe_loss = self.probe_loss(labels, class_logits)
+        gradients = tape.gradient(probe_loss, self.linear_probe.trainable_weights)
+        self.probe_optimizer.apply_gradients(
+            zip(gradients, self.linear_probe.trainable_weights)
+        )
+        self.probe_accuracy.update_state(labels, class_logits)
+
+        return {
+            "c_loss": contrastive_loss,
```

Comment:
If you return a batch-wise scalar here, you will only print the value for the last batch (rather than the average of the value over an epoch, which is standard). Prefer creating 2 loss tracker `keras.metrics.Mean` objects (e.g. in `compile`) and list them in `self.metrics`. Then this line can be simplified as `return {m.name: m.result() for m in self.metrics}`.

See also https://keras.io/guides/customizing_what_happens_in_fit/#going-lowerlevel